keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,509 @@
+- Delete the Pod
+
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled
+- Create `test-pod` again.
+- Verify that the pod is correctly scheduled to the node
+- Remove the node label
+- Verify that the pod is evicted
+- Add the node's label back
+- Create `test-pod` again
+  
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate disabled
+- Verify the old `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` reserved
+- Remove the node label 
+- Verify that the pod is not evicted
+- Add the node's label back 
+
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled
+- Remove the node label
+- Verify that the pod is evicted 
+
+Test2:
+- Start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled.
+- Create a daemonSet with requiredDuringSchedulingRequiredDuringExecution. 
+- Verify that the pod runs only on nodes that meet the nodeAffinity.
+- Remove the node label.
+- Verify that the daemonSet pod on that node is evicted and not recreated.
+- Add the node's label back.
+- Verify that the daemonSet pod runs on the same node again.
+  
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate disabled
+- Verify the old `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` reserved
+- Remove the node label 
+- Verify that the pod is not evicted
+- Add the node's label back 
+
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled
+- Remove the node label
+- Verify that the daemonSet pod on that node is evicted and not recreated.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+- Check the corresponding metrics(details in SLIs part below).
+- Inspect the `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` configuration.
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [x] Events
+  - Event Reason: NodeAffinityEviction
+- [x] API .status
+  - Condition reason: DeletionByNodeAffinityEvictionController 
+- [x] Other (treat as last resort)
+  - Details: Pod with `RequiredDuringSchedulingRequiredDuringExecution` must meet affinity rule during both the scheduling phase and the execution phase.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+N/A
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [x] Metrics
+  - Metric name:
+    - evictions_node_affinity_total(counts of the eviction/evacuation subresources created by the controller)
```

Comment:
> if we can tell the reason here like rejected because PDB

We can't do it currently as both rate limited and violation of pdb will return a 429 status. This depends on the update of the eviction API.

Reply:
Ticked an issue here: https://github.com/kubernetes/kubernetes/issues/125500